kubelet: multiple volumes reference one PVC in one Pod

huww98 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Previously, pod with multiple volumes references one PVC is stuck at ContainerCreating without any error message.
Fixing this by storing multiple OuterVolumeSpecNames per volume.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Pod can have multiple volumes reference the same PVC
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

huww98 committed:
kubelet: multiple volumes reference one PVC in one Pod

Previously, pod with multiple volumes references one PVC is stuck at
ContainerCreating without any error message.
Fixing this by storing multiple OuterVolumeSpecNames per volume

bart0sh commented:
@huww98 this sounds like a bug to me. Please cover this use case with tests, thanks.

/kind bug
/triage accepted
/priority important-soon
/ok-to-test
/cc @gnufied @jsafrane

huww98 commented:
/remove-kind feature

huww98 committed:
test: add assertion to check OuterVolumeSpecNames

Also add a new case for multiple outer names.
And some small fixes to existing cases.

huww98 committed:
e2e: case for multiple volumes reference one PVC

huww98 committed:
remove MountedVolume.OuterVolumeSpecNames from logs

huww98 committed:
kubelet/volumeManager: remove outerVolumeSpecName from ASW

huww98 commented:
Here is what happened. After the rebase, my newly added e2e test failed. I found there is a race condition: ASW synced the middle state of DSW, where only one of two outer names is present. This field has an out-of-sync issue before (#103143), and its fix seems already hacky. So I decide to just remove `outerVolumeSpecName` from DSW.

I have reworked the PR, and organized the changes into different commit logically.

Stage 1: Cleanups that are not directly related to this PR, but are necessary to make CI happy.
* rename Gid => GID
* test: check for error returned by dsw.AddPodToVolume
* fix ExistingPodExistingVolume test case

Stage2: remove dependencies from outerVolumeSpecName in ASW

* kubelet: don't check for mounted before update dsw PV size
* kubelet/volumeManager: verifyVolumesMountedFunc checks both desired and actual
* kubelet/volumeManager: GetMountedVolumesForPod() for desired and actually mounted volumes
* kubelet/volumeManager: (Get -> Has)PossiblyMountedVolumesForPod
* remove MountedVolume.OuterVolumeSpecNames from logs

Stage3: actually remove outerVolumeSpecName from ASW

* kubelet/volumeManager: remove outerVolumeSpecName from ASW

Stage4: support for multiple outerVolumeSpecNames in DSW

* kubelet: multiple volumes reference one PVC in one Pod

Stage5: add tests for this use case

* test: add assertion to check OuterVolumeSpecNames
* e2e: case for multiple volumes reference one PVC

For reviewers: You may review each commit in order.

huww98 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -379,13 +366,6 @@ func (dswp *desiredStateOfWorldPopulator) checkVolumeFSResize(
 		return
 	}
 
-	uniqueVolumeName, exist := getUniqueVolumeName(uniquePodName, podVolume.Name, mountedVolumesForPod)
-	if !exist {
-		// Volume not exist in ASW, we assume it hasn't been mounted yet. If it needs resize,
-		// it will be handled as offline resize(if it indeed hasn't been mounted yet),
-		// or online resize in subsequent loop(after we confirm it has been mounted).
-		return
-	}
```
I think it is safe. The logic of online expansion is still only triggered after the volume is mounted:
https://github.com/kubernetes/kubernetes/blob/0dc369eede901c6027c7de9ad4e6144e6ac8f33d/pkg/kubelet/volumemanager/cache/actual_state_of_world.go#L900-L902

I think this change can avoid an unnecessary resize call if the volume is expanded between attach and mount.

Anyway, determining DSW based on ASW seems awkward to me. It is ASW that should sync from DSW, not the reverse.

jsafrane replied:
I spent some time checking how offline and online resize works and I think you're right.  Offline expansion does not read / use information updated by `UpdatePersistentVolumeSize` / `InitializeClaimSize` below and thus should not be affected by this change.

huww98 commented:
/test pull-kubernetes-e2e-gce

jsafrane commented:
/test pull-kubernetes-e2e-gce-storage-slow

jsafrane commented on pkg/kubelet/volumemanager/volume_manager.go:
```diff
@@ -531,29 +532,35 @@ func (vm *volumeManager) verifyVolumesUnmountedFunc(podName types.UniquePodName)
-// mountedVolumes to a list of unmountedVolumes and returns it.
-func filterUnmountedVolumes(mountedVolumes sets.String, expectedVolumes []string) []string {
+// getUnmountedVolumes returns a list of unmounted volumes.
+// This includes the volumes in expectedVolumes, but not in DSW or ASW.
```
The previous `getUnmountedVolumes` returned volumes that are in `expectedVolumes` and not mounted in ASW. The comment says it filters out DSW volumes, which is IMO not correct.

huww98 replied:
I mean "not in DSW or not in ASW" (not in one of DSW/ASW). Will update the comment.

That is, in addition to the previous `getUnmountedVolumes()`, I regard volumes mounted in ASW but not in DSW as unmounted, which should be rare.

huww98 commented:
```
  [FAILED] deploying csi-hostpath driver: create ClusterRoleBinding: clusterrolebindings.rbac.authorization.k8s.io "psp-csi-hostpath-role-multivolume-5559" already exists
```
I think two test case got the same namespace accidently?

/retest

jsafrane commented:
/lgtm
/approve

huww98 commented:
/assign @dchen1107
for approval. PTAL~

huww98 commented:
rebased and overwrote the comment change in https://github.com/kubernetes/kubernetes/pull/124242/files#diff-952291d712ff384521e6b92c2006c153aa134548beb73f23ac21213e7cb6b932R388-R389

@jsafrane PTAL again, thanks.

jsafrane commented:
/lgtm

huww98 commented:
Resolved trivial conflict to #120844

huww98 commented:
@liggitt Can you take a look at this? Most of this PR already has approval from @jsafrane . We just need another approval for sevaral lines in `pkg/kubelet`. This is an important bug fix, and has lgtm since 2 month ago.

liggitt commented:
> @liggitt Can you take a look at this? Most of this PR already has approval from @jsafrane . We just need another approval for sevaral lines in `pkg/kubelet`. This is an important bug fix, and has lgtm since 2 month ago.

I don't have context on any of the files being changed here... this needs overall approval from someone with very detailed kubelet knowledge, and that is not me :)

huww98 commented:
Resolved trivial conflicts. @jsafrane can you give some advice on how to push this PR forward? It has been waiting for review for 4 months since your approval, without any progress. I have tried to ping some random person, and post this on sig-node slack channel.

gnufied commented:
/assign

jsafrane commented:
/lgtm

gnufied commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -329,7 +315,7 @@ func (dswp *desiredStateOfWorldPopulator) processPodVolumes(
 		}
 
-		dswp.checkVolumeFSResize(pod, podVolume, pvc, volumeSpec, uniquePodName, mountedVolumesForPod)
+		dswp.checkVolumeFSResize(pod, podVolume, pvc, volumeSpec, uniquePodName, uniqueVolumeName)
```
`uniquePodName` is a unused parameter for this function now.

huww98 replied:
Removed this parameter. Thanks

huww98 commented on test/e2e/storage/persistent_volumes.go:
```diff
@@ -211,6 +211,25 @@ var _ = utils.SIGDescribe("PersistentVolumes", func() {
 			})
 
+			// The same as above, but with multiple volumes reference the same PVC in the pod.
+			ginkgo.It("create a PVC and use it multiple times in a single pod", func(ctx context.Context) {
```
Yes, it works! I just tested this case manually.

huww98 commented:
/retest

liggitt commented:
Does this change the disk structure of pod volumes in a way that will break a kubelet that does a minor upgrade without draining pods with volumes?

huww98 commented:
> Does this change the disk structure of pod volumes in a way that will break a kubelet that does a minor upgrade without draining pods with volumes?

No, this PR should only change the in-memory data structure. No breaking is expected.

huww98 commented:
Resolved trivial conflicts, dropped the changes already merged in #125139

huww98 commented:
#126575 added another usage to ASW outerVolumeSpecName. I've read that PR, and implement the logic in another way. Specifically, the following commits are rewritten based on his idea, but avoids using ASW outerVolumeSpecName. cc @Lucaber
* kubelet/volumeManager: verifyVolumesMountedFunc checks both desired and actual
* kubelet/volumeManager: GetMountedVolumesForPod() for desired and actually mounted volumes

huww98 commented:
Dropped 3 commits already merged in #128522 . I will continue to split this large PR to more smaller ones.

huww98 commented:
/test pull-kubernetes-e2e-gce

huww98 commented:
Rebased after #128649 merged. We are one more step closer!
@jsafrane PTAL again, thanks. This should be fine since there is no conflicts when rebasing.

jsafrane commented:
/lgtm
